src/cirq_iqm/iqm_client.py

```diff
@@ -0,0 +1,179 @@
+# Copyright 2020–2021 Cirq on IQM developers
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+Implements a client for calling the IQM backend
+"""
+import time
+from dataclasses import dataclass
+import json
+from enum import Enum
+from datetime import datetime
+from typing import Any, Union
+
+import requests
+
+TIMEOUT_SECONDS = 10
+SECONDS_BETWEEN_CALLS = 1
+
+
+class CircuitExecutionException(Exception):
+    """
+    Something went wrong on the server side
+    """
+
+
+class ApiTimeoutError(CircuitExecutionException):
+    """
+    Exception for when executing a task on the backend takes too long
+    """
+
+
+class RunStatus(str, Enum):
+    """
+    Status of a task
+    """
+    PENDING = "pending"
+    READY = "ready"
+    FAILED = "failed"
+
+
+@dataclass(frozen=True)
+class IQMInstruction:
+    """
+    Transfer DTO for IQM insrtructions
+    """
+    name: str
+    qubits: list[str]
+    args: dict[str, Any]
+
+
+@dataclass(frozen=True)
+class IQMCircuit:
+    """
+    Transfer DTO for IQM circuit
+    """
+    name: str
+    args: dict
+    instructions: list[IQMInstruction]
+
+
+@dataclass(frozen=True)
+class QubitMapping:
+    """
+    Mapping of logical qubits to physical qubits
+    """
+    logical_name: str
+    physical_name: str
+
+
+@dataclass(frozen=True)
+class RunResult:
+    """
+    Result of a task execution
+    """
+    status: RunStatus
+    measurements: dict[str:list[list]] = None
+    message: str = None
```

Comment:
Isn't it so that a RunStatus has either a `measurements` or `message`? I don't know if there is a good way to code that in typing but there could at least be a comment about that.

Reply:
Initially, I had several types for this however, it makes a cleaner python code. 
This is how requests.Response object works. Even if it is 200, it has a "reason" field.
Parsing a complex object would require more code in python and we probably don't want to maintain our own parsers and 3 classes of Results just to drop "message"

Reply:
Yes, that's why I'm fine with this typing but the docstring should probably explain that in the real server response, the message field is only present for an error case and the measurements for a success and they can't both be present.

Reply:
sure